File: kamon-executors/src/main/java/kamon/executors/instrumentation/ExecutorInstrumentation.java
/* =========================================================================================
 * Copyright © 2013-2018 the kamon project <http://kamon.io/>
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 * =========================================================================================
 */

package kamon.executors.instrumentation;

import kamon.Kamon;
import kamon.context.Context;
import kamon.context.Storage.Scope;
import kamon.executors.instrumentation.ExecutorsInstrumentationAdvisors.CallableCollectionWrapperAdvisor;
import kamon.executors.instrumentation.ExecutorsInstrumentationAdvisors.CallableWrapperAdvisor;
import kamon.executors.instrumentation.ExecutorsInstrumentationAdvisors.RunnableWrapperAdvisor;
import kanela.agent.api.instrumentation.KanelaInstrumentation;
import kanela.agent.bootstrap.context.ContextHandler;
import kanela.agent.bootstrap.context.ContextProvider;

import java.util.Collection;
import java.util.concurrent.Callable;

public final class ExecutorInstrumentation extends KanelaInstrumentation {

    public ExecutorInstrumentation() {
        /**
         * Set the ContextProvider
         */
        ContextHandler.setContexProvider(new KamonContextProvider());

        /**
         * Instrument all implementations of:
         *
         * java.util.concurrent.Executor::execute
         *
         */
        forSubtypeOf(() -> "java.util.concurrent.Executor", builder ->
                builder
                    .withAdvisorFor(method("execute").and(withArgument(Runnable.class)), () -> RunnableWrapperAdvisor.class)
                    .build());


        /**
         * Instrument all implementations of:
         *
         * java.util.concurrent.ExecutorService::submit(Runnable)
         * java.util.concurrent.ExecutorService::submit(Callable)
         * java.util.concurrent.ExecutorService::[invokeAny|invokeAll](Collection[Callable])
         *
         */
        forSubtypeOf(() -> "java.util.concurrent.ExecutorService", builder ->
                builder
                    .withAdvisorFor(method("submit").and(withArgument(Runnable.class)), () -> RunnableWrapperAdvisor.class)
                    .withAdvisorFor(method("submit").and(withArgument(Callable.class)), () -> CallableWrapperAdvisor.class)
                    .withAdvisorFor(anyMethods("invokeAny", "invokeAll").and(withArgument(Collection.class)), () -> CallableCollectionWrapperAdvisor.class)
                    .build());
    }

    /**
     * Runs a Runnable within Kamon Context
     */
    private static class ContextAwareRunnable implements Runnable {

        private final Runnable underlying;
        private final Context context;

        ContextAwareRunnable(Runnable r) {
            this.context = Kamon.currentContext();
            this.underlying = r;
        }

        @Override
        public void run() {
            final Scope scope = Kamon.storeContext(context);
            try {
                underlying.run();
            } finally {
                scope.close();
            }
        }
    }

    /**
     * Runs a Callable within Kamon Context
     */
    private static class ContextAwareCallable<A> implements Callable<A> {

        private final Callable<A> underlying;
        private final Context context;

        ContextAwareCallable(Callable<A> c) {
            this.context = Kamon.currentContext();
            this.underlying = c;
        }

        public A call() throws Exception {
            final Scope scope = Kamon.storeContext(context);
            try {
                return underlying.call();
            } finally {
                scope.close();
            }
        }
    }

    /**
     * implementation of kanela.agent.bootstrap.context.ContextProvider
     */
    private static class KamonContextProvider implements ContextProvider {
        @Override
        public Runnable wrapInContextAware(Runnable runnable) {
            return new ContextAwareRunnable(runnable);
        }

        @Override
        public <A> Callable wrapInContextAware(Callable<A> callable) {
            return new ContextAwareCallable<>(callable);
        }
    }
}

File: instrumentation/cassandra/src/main/scala/kamon/instrumentation/cassandra/executors/ExecutorInstrumentation.scala
/*
 * Copyright 2013-2021 The Kamon Project <https://kamon.io>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package kamon.instrumentation.cassandra.executors

import java.util.concurrent.{Callable, ExecutorService, ScheduledExecutorService}
import kamon.instrumentation.cassandra.CassandraInstrumentation
import kamon.instrumentation.executor.ExecutorInstrumentation
import kamon.tag.TagSet
import kanela.agent.api.instrumentation.InstrumentationBuilder
import kanela.agent.libs.net.bytebuddy.implementation.bind.annotation.SuperCall

import scala.annotation.static

class DriverExecutorInstrumentation extends InstrumentationBuilder {

  /**
    * Wraps all executors created by the Cassandra driver with the Kamon executors instrumentation.
    */
  onType("com.datastax.driver.core.ThreadingOptions")
    .intercept(method("createExecutor"), classOf[CreateExecutorAdvice])
    .intercept(method("createBlockingExecutor"), classOf[CreateBlockingTasksExecutorAdvice])
    .intercept(method("createReaperExecutor"), classOf[CreateReaperExecutorAdvice])
    .intercept(method("createScheduledTasksExecutor"), classOf[CreateScheduledTasksExecutorAdvice])
    .intercept(method("createReconnectionExecutor"), classOf[CreateReconnectionExecutorAdvice])
}

sealed trait ExecutorMetrics {

  val componentTags = TagSet.of("component", CassandraInstrumentation.Tags.CassandraDriverComponent)

  def metricName(executorName: String) =
    "cassandra.driver.executor." + executorName

  def instrument(callable: Callable[ExecutorService], name: String): ExecutorService =
    ExecutorInstrumentation.instrument(callable.call(), metricName(name), componentTags)

  def instrumentScheduled(callable: Callable[ScheduledExecutorService], name: String): ScheduledExecutorService =
    ExecutorInstrumentation.instrumentScheduledExecutor(callable.call(), metricName(name), componentTags)
}

class CreateExecutorAdvice
object CreateExecutorAdvice extends ExecutorMetrics {
  @static def onExecutorCreated(@SuperCall callable: Callable[ExecutorService]): ExecutorService =
    instrument(callable, "executor")
}

class CreateBlockingTasksExecutorAdvice
object CreateBlockingTasksExecutorAdvice extends ExecutorMetrics {
  @static def onExecutorCreated(@SuperCall callable: Callable[ExecutorService]): ExecutorService =
    instrument(callable, "blocking")
}

class CreateReaperExecutorAdvice
object CreateReaperExecutorAdvice extends ExecutorMetrics {
  @static def onExecutorCreated(
    @SuperCall callable: Callable[ScheduledExecutorService]
  ): ScheduledExecutorService =
    instrumentScheduled(callable, "reaper")
}

class CreateScheduledTasksExecutorAdvice
object CreateScheduledTasksExecutorAdvice extends ExecutorMetrics {
  @static def onExecutorCreated(
    @SuperCall callable: Callable[ScheduledExecutorService]
  ): ScheduledExecutorService =
    instrumentScheduled(callable, "scheduled-tasks")
}

class CreateReconnectionExecutorAdvice
object CreateReconnectionExecutorAdvice extends ExecutorMetrics {
  @static def onExecutorCreated(
    @SuperCall callable: Callable[ScheduledExecutorService]
  ): ScheduledExecutorService =
    instrumentScheduled(callable, "reconnection")
}
